d2graph/d2graph.go

```diff
@@ -898,166 +1056,66 @@ func appendTextDedup(texts []*d2target.MText, t *d2target.MText) []*d2target.MTe
 			// no labels
 		default:
 			if obj.Attributes.Label.Value != "" {
-				obj.LabelWidth = go2.Pointer(dims.Width)
-				obj.LabelHeight = go2.Pointer(dims.Height)
+				obj.LabelWidth = go2.Pointer(labelDims.Width)
+				obj.LabelHeight = go2.Pointer(labelDims.Height)
 			}
 		}
 
-		dims.Width += innerLabelPadding
-		dims.Height += innerLabelPadding
-		obj.LabelDimensions = *dims
-		obj.Width = float64(dims.Width)
-		obj.Height = float64(dims.Height)
-
-		switch strings.ToLower(obj.Attributes.Shape.Value) {
-		default:
-			obj.Width += 100
-			obj.Height += 100
-
-		case d2target.ShapeImage:
-			if obj.Attributes.Width != nil {
-				w, _ := strconv.Atoi(obj.Attributes.Width.Value)
-				obj.Width = float64(w)
-			} else {
-				obj.Width = 128
-			}
-			if obj.Attributes.Height != nil {
-				h, _ := strconv.Atoi(obj.Attributes.Height.Value)
-				obj.Height = float64(h)
-			} else {
-				obj.Height = 128
-			}
+		if shapeType != d2target.ShapeText && obj.Attributes.Label.Value != "" {
+			labelDims.Width += INNER_LABEL_PADDING
+			labelDims.Height += INNER_LABEL_PADDING
+		}
+		obj.LabelDimensions = *labelDims
 
-		case d2target.ShapeSquare, d2target.ShapeCircle:
-			sideLength := go2.Max(obj.Width, obj.Height)
-			obj.Width = sideLength + 100
-			obj.Height = sideLength + 100
+		defaultDims, err := obj.GetDefaultSize(mtexts, ruler, fontFamily, *labelDims)
+		if err != nil {
+			return err
+		}
 
-		case d2target.ShapeClass:
-			maxWidth := dims.Width
+		obj.Width = float64(go2.Max(defaultDims.Width, desiredWidth))
+		obj.Height = float64(go2.Max(defaultDims.Height, desiredHeight))
 
-			for _, f := range obj.Class.Fields {
-				fdims := GetTextDimensions(mtexts, ruler, f.Text(), go2.Pointer(d2fonts.SourceCodePro))
-				if fdims == nil {
-					return fmt.Errorf("dimensions for class field %#v not found", f.Text())
-				}
-				lineWidth := fdims.Width
-				if maxWidth < lineWidth {
-					maxWidth = lineWidth
-				}
-			}
-			for _, m := range obj.Class.Methods {
-				mdims := GetTextDimensions(mtexts, ruler, m.Text(), go2.Pointer(d2fonts.SourceCodePro))
-				if mdims == nil {
-					return fmt.Errorf("dimensions for class method %#v not found", m.Text())
-				}
-				lineWidth := mdims.Width
-				if maxWidth < lineWidth {
-					maxWidth = lineWidth
-				}
-			}
+		paddingX, paddingY := obj.GetPadding()
 
-			// All rows should be the same height
-			var anyRowText *d2target.MText
-			if len(obj.Class.Fields) > 0 {
-				anyRowText = obj.Class.Fields[0].Text()
-			} else if len(obj.Class.Methods) > 0 {
-				anyRowText = obj.Class.Methods[0].Text()
-			}
-			if anyRowText != nil {
-				// 10px of padding top and bottom so text doesn't look squished
-				rowHeight := GetTextDimensions(mtexts, ruler, anyRowText, go2.Pointer(d2fonts.SourceCodePro)).Height + 20
-				obj.Height = float64(rowHeight * (len(obj.Class.Fields) + len(obj.Class.Methods) + 2))
+		switch shapeType {
+		case d2target.ShapeSquare, d2target.ShapeCircle:
```

Comment:
let's make this a compiler error if they specify different values for width/height for these shapes